internal/gatewayapi/listener.go

```diff
@@ -172,9 +176,172 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR resource
+			if listener.httpIR != nil {
+				listener.httpIR.TLSOverlaps = true
+			}
+		}
 	}
 }
 
+type overlappingListener struct {
+	listener1 string
+	listener2 string
+	san1      string
+	san2      string
+}
+
+func checkOverlappingCertificates(gateway *GatewayContext) {
+	httpsListeners := []*ListenerContext{}
+	for _, listener := range gateway.listeners {
+		if listener.Protocol == gwapiv1.HTTPSProtocolType {
+			httpsListeners = append(httpsListeners, listener)
+		}
+	}
+
+	overlappingListeners := make([]*overlappingListener, len(httpsListeners)) // The listener name and san of the overlapping listener.
+	for i := range httpsListeners {
+		if overlappingListeners[i] != nil {
+			continue
+		}
+		for j := i + 1; j < len(httpsListeners); j++ {
+			if overlappingListeners[j] != nil {
+				continue
+			}
+			if httpsListeners[i].Port != httpsListeners[j].Port {
+				continue
+			}
+			for _, cert1 := range httpsListeners[i].tlsCertificates {
+				for _, cert2 := range httpsListeners[j].tlsCertificates {
+					overlappingCertificate := isOverlappingCertificate(cert1, cert2)
+					if overlappingCertificate != nil {
+						// Overlapping listeners can be more than two, we only report the first two for simplicity.
+						overlappingListeners[i] = &overlappingListener{
+							listener1: string(httpsListeners[i].Name),
+							listener2: string(httpsListeners[j].Name),
+							san1:      overlappingCertificate.san1,
+							san2:      overlappingCertificate.san2,
+						}
+						overlappingListeners[j] = &overlappingListener{
+							listener1: string(httpsListeners[j].Name),
+							listener2: string(httpsListeners[i].Name),
+							san1:      overlappingCertificate.san2,
+							san2:      overlappingCertificate.san1,
+						}
+					}
+				}
+			}
+		}
+	}
+
+	for i, listener := range httpsListeners {
+		if overlappingListeners[i] != nil {
+			status.SetGatewayListenerStatusCondition(gateway.Gateway,
+				listener.listenerStatusIdx,
+				gwapiv1.ListenerConditionOverlappingTLSConfig,
+				metav1.ConditionTrue,
+				gwapiv1.ListenerReasonOverlappingCertificates,
+				fmt.Sprintf(
+					"The certificate san %s overlaps with the certificate san %s in listener %s",
+					overlappingListeners[i].san1,
+					overlappingListeners[i].san2,
+					overlappingListeners[i].listener2,
+				),
+			)
+			if listener.httpIR != nil {
+				listener.httpIR.TLSOverlaps = true
+			}
+		}
+	}
+}
+
+type overlappingCertificate struct {
+	san1 string
+	san2 string
+}
+
+func isOverlappingCertificate(cert1, cert2 *x509.Certificate) *overlappingCertificate {
+	for _, dns1 := range cert1.DNSNames {
+		for _, dns2 := range cert2.DNSNames {
+			if isOverlappingHostname(ptr.To(gwapiv1.Hostname(dns1)), ptr.To(gwapiv1.Hostname(dns2))) {
+				return &overlappingCertificate{
+					san1: dns1,
+					san2: dns2,
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// isOverlappingHostname checks if two hostnames overlap.
+func isOverlappingHostname(hostname1, hostname2 *gwapiv1.Hostname) bool {
```

Comment:
can we reuse / dedup some of https://github.com/envoyproxy/gateway/blob/7b64768859d1221c039fe0e767fbf1bc41b42e5c/internal/gatewayapi/route.go#L820 ?

Reply:
I don't think we can. These are handling two different logics.